test/e2e/vmagent_test.go

```diff
@@ -291,3 +302,411 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
+	anpRule := builder.Get()
+	i := 0
+	for fqdn, action := range fqdnSettings {
+		ruleName := fmt.Sprintf("name-%d", i)
+		policyPeer := []crdv1alpha1.NetworkPolicyPeer{{FQDN: fqdn}}
+		ports, _ := GenPortsOrProtocols(ProtocolTCP, nil, nil, nil, nil, nil, nil, nil)
+		newRule := crdv1alpha1.Rule{
+			To:     policyPeer,
+			Ports:  ports,
+			Action: action,
+			Name:   ruleName,
+		}
+		anpRule.Spec.Egress = append(anpRule.Spec.Egress, newRule)
+		i += 1
+	}
+
+	anp, err := data.CreateOrUpdateANP(anpRule)
+	require.Nil(t, err, "Failed to create Antrea NetworkPolicy")
+	assert.Nil(t, data.waitForANPRealized(t, anp.Namespace, anp.Name, policyRealizedTimeout), "Failed to realize Antrea NetworkPolicy")
+	return anp
+}
+
+func verifyANPConnectivity(t *testing.T, data *TestData, srcVM vmInfo, dstVM vmInfo, connected bool, proto AntreaPolicyProtocol, ruleAction crdv1alpha1.RuleAction) error {
+	var err error
+	switch proto {
+	case ProtocolTCP:
+		fallthrough
+	case ProtocolUDP:
+		err = runIperfCommandOnVMs(t, data, srcVM, dstVM, connected, proto == ProtocolUDP, ruleAction)
+	case ProtocolICMP:
+		err = runPingCommandOnVM(t, data, dstVM, connected)
+	}
+	return err
+}
+
+func runPingCommandOnVM(t *testing.T, data *TestData, dstVM vmInfo, connected bool) error {
+	dstIP := net.ParseIP(dstVM.ip)
+	cmd := getPingCommand(pingCount, 0, strings.ToLower(linuxOS), &dstIP)
+	cmdStr := strings.Join(cmd, " ")
+	expCount := pingCount
+	if !connected {
+		expCount = 0
+	}
+	expOutput := fmt.Sprintf("%d packets transmitted, %d received", pingCount, expCount)
+	// Use master Node to run ping command.
+	pingClient := nodeName(0)
+	err := wait.PollImmediate(time.Second*5, time.Second*20, func() (done bool, err error) {
+		if err := runCommandAndCheckResult(data, pingClient, cmdStr, expOutput, ""); err != nil {
+			t.Logf("Failed to run ping command on VM %s: %v", pingClient, err)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func runIperfCommandOnVMs(t *testing.T, data *TestData, srcVM vmInfo, dstVM vmInfo, connected bool, isUDP bool, ruleAction crdv1alpha1.RuleAction) error {
+	svrIP := net.ParseIP(dstVM.ip)
+	err := wait.PollImmediate(time.Second*5, time.Second*20, func() (done bool, err error) {
+		if err := runIperfClient(t, data, srcVM, svrIP, iperfPort, isUDP, connected, ruleAction); err != nil {
+			t.Errorf("Failed to run iperf asR client, retrying: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func runIperfServer(t *testing.T, data *TestData, vm vmInfo, dstPort int32) error {
+	cmd := getIperf3Command(vm.osType, nil, dstPort, false, true)
+	cmdStr := strings.Join(cmd, " ")
+	if vm.osType == windowsOS {
+		cmdStr = fmt.Sprintf(`cmd.exe /c "%s"`, cmdStr)
+	}
+	_, _, _, err := data.provider.RunCommandOnNode(vm.nodeName, cmdStr)
+	if err != nil {
+		t.Errorf("Failed to run iperf3 command on VM %s, err %v", vm.nodeName, err)
+		return err
```

Comment:
Similar to non test code, this would lead to [duplicate error logs](https://github.com/tnqn/code-review-comments#do-not-handle-errors-repeatedly), I think you logged exactly same message in `testANPWithConnectivity`. This comment applies to several places.